File: tests/tSiNCs/gen_nanocrystals.py
#!/usr/bin/env python3
"""Nanocrystal generator — Python CLI. Sphere cut: native; Miller planes + bridges: node tests/tSiNCs/gen_nanocrystals.mjs."""
import argparse
import os
import subprocess
import sys

REPO = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
TEST_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, REPO)

from pyBall.nanocrystal_gen import build_spherical_nanoparticle, save_xyz, find_cap_hh_pairs, save_hh_bonds_json


def run_node_planes(args):
    cmd = ['node', os.path.join(TEST_DIR, 'gen_nanocrystals.mjs'),
           '--samples', '1', '--maxFiles', '1',
           '--cutMode', 'planes',
           '--cif', args.cif,
           '--caps', args.caps,
           '--outDir', args.out_dir,
           '--prefix', args.prefix,
           '--bondFactor', str(args.bond_factor),
           '--defaultRcut', str(args.default_rcut),
           '--minHeavyDegree', str(args.min_heavy_degree),
           '--pruneMaxIter', str(args.prune_max_iter),
           '--outwardBias', str(args.outward_bias),
           '--insertProb', str(args.insert_prob),
           '--collapseProb', str(args.collapse_prob),
           '--nx-range', args.nx_range,
           '--ny-range', args.ny_range,
           '--nz-range', args.nz_range,
           '--centered', '1' if args.centered else '0',
           '--planeTemplates', args.plane_templates,
           '--planeSymC', str(args.plane_sym_c),
           '--planeCScale', str(args.plane_c_scale),
           '--planeCJitter', str(args.plane_c_jitter),
           ]
    if args.cap_hh_bonds:
        cmd.extend(['--capHHBonds', '1', '--capHHBondDist', str(args.cap_hh_bond_dist)])
    if args.seed:
        cmd.extend(['--seed', str(args.seed)])
    print('[gen_nanocrystals.py] node', ' '.join(cmd[2:]))
    subprocess.run(cmd, cwd=REPO, check=True)


def run_sphere_native(args):
    heavy_z = 14 if args.element.lower() == 'si' else 6
    prim = args.primitive_xyz
    if not prim:
        prim = os.path.join(REPO, 'cpp', 'common_resources', 'crystals',
                            'Si_primitive.xyz' if heavy_z == 14 else 'diamond_primitive.xyz')
    elems, apos = build_spherical_nanoparticle(
        prim_xyz=prim, R=args.sphere_r, nrep=args.sphere_nrep,
        min_degree=args.min_heavy_degree, heavy_z=heavy_z,
        rcut_heavy=args.rcut_heavy, outward_bias=args.outward_bias,
        resolve_clashes=(args.caps == 'H'),
    )
    os.makedirs(args.out_dir, exist_ok=True)
    tag = f'{args.prefix}_sphere_R{args.sphere_r:.1f}_nrep{args.sphere_nrep}_nat{len(elems)}'
    xyz_path = os.path.join(args.out_dir, tag + '.xyz')
    save_xyz(xyz_path, elems, apos, comment=f'sphere R={args.sphere_r} nrep={args.sphere_nrep} element={args.element}')
    print(f'[gen_nanocrystals.py] wrote {xyz_path} atoms={len(elems)}')
    if args.cap_hh_bonds:
        pairs = find_cap_hh_pairs(apos, elems, args.cap_hh_bond_dist)
        bonds_path = xyz_path + '.hh_bonds.json'
        save_hh_bonds_json(bonds_path, pairs)
        print(f'[gen_nanocrystals.py] wrote {bonds_path} hhBonds={len(pairs)}')
    if args.insert_prob > 0 or args.collapse_prob > 0:
        print('[gen_nanocrystals.py] WARNING: bridge insert/collapse requires --cutMode planes (node engine)')
    return xyz_path


def main():
    ap = argparse.ArgumentParser(description='Nanocrystal generator (Python)')
    ap.add_argument('--cutMode', choices=['sphere', 'planes'], default='sphere')
    ap.add_argument('--element', choices=['C', 'Si', 'c', 'si'], default='C')
    ap.add_argument('--cif', default=os.path.join(REPO, 'cpp/common_resources/crystals/Si-sym.cif'))
    ap.add_argument('--primitive-xyz', default=None)
    ap.add_argument('--sphere-r', type=float, default=6.0, dest='sphere_r')
    ap.add_argument('--sphere-nrep', type=int, default=5, dest='sphere_nrep')
    ap.add_argument('--rcut-heavy', type=float, default=1.75, dest='rcut_heavy')
    ap.add_argument('--caps', default='H')
    ap.add_argument('--outDir', dest='out_dir', default=os.path.join(TEST_DIR, 'OUT_nanocrystals_py'))
    ap.add_argument('--prefix', default='nc_py')
    ap.add_argument('--bondFactor', type=float, default=1.30, dest='bond_factor')
    ap.add_argument('--defaultRcut', type=float, default=1.60, dest='default_rcut')
    ap.add_argument('--minHeavyDegree', type=int, default=2, dest='min_heavy_degree')
    ap.add_argument('--pruneMaxIter', type=int, default=10, dest='prune_max_iter')
    ap.add_argument('--outwardBias', type=float, default=0.35, dest='outward_bias')
    ap.add_argument('--insertProb', type=float, default=0.0, dest='insert_prob')
    ap.add_argument('--collapseProb', type=float, default=0.0, dest='collapse_prob')
    ap.add_argument('--seed', type=int, default=0)
    ap.add_argument('--nx-range', default='2,2', dest='nx_range')
    ap.add_argument('--ny-range', default='2,2', dest='ny_range')
    ap.add_argument('--nz-range', default='2,2', dest='nz_range')
    ap.add_argument('--centered', action='store_true', default=True)
    ap.add_argument('--planeTemplates', default='a111', dest='plane_templates')
    ap.add_argument('--planeSymC', type=float, default=6.0, dest='plane_sym_c')
    ap.add_argument('--planeCScale', type=float, default=0.40, dest='plane_c_scale')
    ap.add_argument('--planeCJitter', type=float, default=0.0, dest='plane_c_jitter')
    ap.add_argument('--capHHBonds', action='store_true', dest='cap_hh_bonds')
    ap.add_argument('--capHHBondDist', type=float, default=1.8, dest='cap_hh_bond_dist')
    args = ap.parse_args()

    if args.cutMode == 'planes':
        run_node_planes(args)
    else:
        run_sphere_native(args)


if __name__ == '__main__':
    main()

File: tests/tSiNCs/gen_nanocrystals.mjs
/// @deprecated Use `node tests/tSiNCs/nanocrystals.mjs generate` instead. This script is kept for backward compatibility.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { toMol2String, toXYZString } from '../../web/molgui_webgpu/MoleculeIO.js';
import {
    loadMMParams, buildCrystalFromCIFText, getHeavyZ, generateNanocrystal, mulberry32,
} from '../../web/molgui_webgpu/Nanocrystals.js';

function parseProb(s, name) {
    const v = +String(s).trim();
    if (!isFinite(v) || v < 0 || v > 1) throw new Error(`${name}: expected probability in [0,1], got '${s}'`);
    return v;
}

function parseCSVRange(s) {
    const t = String(s || '').trim();
    if (!t) throw new Error('parseCSVRange: empty');
    const parts = t.split(',').map(x => x.trim()).filter(x => x.length > 0);
    if (parts.length === 1) {
        const v = parseInt(parts[0], 10);
        if (!isFinite(v)) throw new Error(`parseCSVRange: bad int '${s}'`);
        return [v | 0, v | 0];
    }
    if (parts.length === 2) {
        const a = parseInt(parts[0], 10);
        const b = parseInt(parts[1], 10);
        if (!isFinite(a) || !isFinite(b)) throw new Error(`parseCSVRange: bad range '${s}'`);
        return [a | 0, b | 0];
    }
    throw new Error(`parseCSVRange: expected 'n' or 'a,b', got '${s}'`);
}

function parsePlaneTemplates(s) {
    const t = String(s || '').trim();
    if (!t) return [];
    return t.split(',').map(x => x.trim()).filter(x => x.length > 0);
}

function parsePlaneSymC(s) {
    const v = +String(s || '').trim();
    if (!isFinite(v) || !(v > 0)) throw new Error(`bad --planeSymC '${s}'`);
    return v;
}

function parseArgs(argv) {
    const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TEST_DIR = __dirname;
const repoRoot = path.resolve(__dirname, '../..');
    const defaultRes = path.resolve(repoRoot, 'cpp/common_resources');

    const out = {
        cif: path.resolve(defaultRes, 'crystals/Si-sym.cif'),
        applySymmetry: true,
        dedupSymmetry: true,
        dedupTol: 0.1,

        samples: 0,
        nx: [3, 3],
        ny: [3, 3],
        nz: [3, 3],
        centered: true,

        planeTemplates: ['a111'],
        planeSymC: 6.0,
        planeMode: 'ang',

        planeC0: 0.0,
        planeCScale: 0.55,
        planeCJitter: 0.10,

        bonds: true,
        elementTypes: path.join(defaultRes, 'ElementTypes.dat'),
        atomTypes: path.join(defaultRes, 'AtomTypes.dat'),
        bondTypes: path.join(defaultRes, 'BondTypes.dat'),
        angleTypes: path.join(defaultRes, 'AngleTypes.dat'),
        bondFactor: 1.30,
        defaultRcut: 1.60,

        caps: 'H',
        minSiDegree: 2,
        pruneMaxIter: 10,

        collapse: 0,
        collapseAll: false,
        requireH2: true,

        collapseProb: 0.0,
        insertProb: 0.0,
        fuseProb: 0.0,
        fuseHClashMax: 2.0,
        outwardBias: 0.35,
        resolveClashes: 1,
        capHHBonds: 0,
        capHHBondDist: 1.8,

        cutMode: 'planes',
        sphereR: 6.0,
        sphereNrep: 5,
        rcutHeavy: 0.0,
        minHeavyDegree: 2,

        E_SiH2: 1.0,
        E_SiH3: 4.0,
        E_bare: 10.0,
        E_bridge: 2.0,
        muH: 0.0,

        outDir: path.resolve(TEST_DIR, 'OUT_nanocrystals'),
        prefix: 'si_nc',
        maxFiles: 0,
        seed: 0,
        statsCsv: null,
        stackedXyzOut: null,
    };

    for (let i = 2; i < argv.length; i++) {
        const a = String(argv[i]);
        const nxt = () => {
            if (i + 1 >= argv.length) throw new Error(`Missing value after ${a}`);
            i++;
            return String(argv[i]);
        };

        if (a === '--cif') out.cif = path.resolve(nxt());
        else if (a === '--applySymmetry') out.applySymmetry = (nxt() !== '0');
        else if (a === '--dedupSymmetry') out.dedupSymmetry = (nxt() !== '0');
        else if (a === '--dedupTol') out.dedupTol = +nxt();

        else if (a === '--samples') out.samples = parseInt(nxt(), 10) | 0;
        else if (a === '--nx' || a === '--nx-range') out.nx = parseCSVRange(nxt());
        else if (a === '--ny' || a === '--ny-range') out.ny = parseCSVRange(nxt());
        else if (a === '--nz' || a === '--nz-range') out.nz = parseCSVRange(nxt());
        else if (a === '--centered') out.centered = (nxt() !== '0');

        else if (a === '--planeTemplates') out.planeTemplates = parsePlaneTemplates(nxt());
        else if (a === '--planeSymC') out.planeSymC = parsePlaneSymC(nxt());
        else if (a === '--planeMode') out.planeMode = nxt();

        else if (a === '--planeC0') out.planeC0 = +nxt();
        else if (a === '--planeCScale') out.planeCScale = +nxt();
        else if (a === '--planeCJitter') out.planeCJitter = +nxt();

        else if (a === '--bonds') out.bonds = (nxt() !== '0');
        else if (a === '--elementTypes') out.elementTypes = path.resolve(nxt());
        else if (a === '--atomTypes') out.atomTypes = path.resolve(nxt());
        else if (a === '--bondTypes') out.bondTypes = path.resolve(nxt());
        else if (a === '--angleTypes') out.angleTypes = path.resolve(nxt());
        else if (a === '--bondFactor') out.bondFactor = +nxt();
        else if (a === '--defaultRcut') out.defaultRcut = +nxt();

        else if (a === '--caps') out.caps = nxt();
        else if (a === '--minSiDegree') out.minHeavyDegree = parseInt(nxt(), 10) | 0;
        else if (a === '--pruneMaxIter') out.pruneMaxIter = parseInt(nxt(), 10) | 0;

        else if (a === '--collapse') out.collapse = parseInt(nxt(), 10) | 0;
        else if (a === '--collapseAll') out.collapseAll = (nxt() !== '0');
        else if (a === '--requireH2') out.requireH2 = (nxt() !== '0');

        else if (a === '--collapseProb') out.collapseProb = parseProb(nxt(), '--collapseProb');
        else if (a === '--insertProb') out.insertProb = parseProb(nxt(), '--insertProb');
        else if (a === '--fuseProb') out.fuseProb = parseProb(nxt(), '--fuseProb');
        else if (a === '--fuseHClashMax') out.fuseHClashMax = +nxt();
        else if (a === '--outwardBias') out.outwardBias = +nxt();
        else if (a === '--resolveClashes') out.resolveClashes = (nxt() !== '0');
        else if (a === '--capHHBonds') out.capHHBonds = (nxt() !== '0');
        else if (a === '--capHHBondDist') out.capHHBondDist = +nxt();

        else if (a === '--cutMode') out.cutMode = nxt();
        else if (a === '--sphereR' || a === '--sphere-r') out.sphereR = +nxt();
        else if (a === '--sphereNrep' || a === '--sphere-nrep') out.sphereNrep = parseInt(nxt(), 10) | 0;
        else if (a === '--rcutHeavy' || a === '--rcut-heavy') out.rcutHeavy = +nxt();
        else if (a === '--minHeavyDegree') out.minHeavyDegree = parseInt(nxt(), 10) | 0;

        else if (a === '--E_SiH2') out.E_SiH2 = +nxt();
        else if (a === '--E_SiH3') out.E_SiH3 = +nxt();
        else if (a === '--E_bare') out.E_bare = +nxt();
        else if (a === '--E_bridge') out.E_bridge = +nxt();
        else if (a === '--muH') out.muH = +nxt();

        else if (a === '--outDir') out.outDir = path.resolve(nxt());
        else if (a === '--prefix') out.prefix = nxt();
        else if (a === '--maxFiles') out.maxFiles = parseInt(nxt(), 10) | 0;
        else if (a === '--seed') out.seed = parseInt(nxt(), 10) | 0;

        else if (a === '--statsCsv') out.statsCsv = path.resolve(nxt());
        else if (a === '--stackedXyzOut') out.stackedXyzOut = path.resolve(nxt());

        else throw new Error(`Unknown arg ${a}`);
    }

    if (out.planeMode !== 'ang' && out.planeMode !== 'frac') throw new Error("--planeMode must be 'ang' or 'frac'");
    if (out.cutMode === 'planes' && (!Array.isArray(out.planeTemplates) || out.planeTemplates.length === 0)) throw new Error('--planeTemplates must be non-empty for cutMode=planes');
    if (!(out.dedupTol > 0)) throw new Error('--dedupTol must be >0');
    if (out.collapse < 0) throw new Error('--collapse must be >=0');
    if (out.samples < 0) throw new Error('--samples must be >=0');
    if (out.minSiDegree < 0) throw new Error('--minSiDegree must be >=0');
    if (out.pruneMaxIter < 0) throw new Error('--pruneMaxIter must be >=0');
    if (!(isFinite(out.planeCScale) && out.planeCScale >= 0)) throw new Error('--planeCScale must be >=0');
    if (!(isFinite(out.planeCJitter) && out.planeCJitter >= 0)) throw new Error('--planeCJitter must be >=0');
    if (!(isFinite(out.capHHBondDist) && out.capHHBondDist > 0)) throw new Error('--capHHBondDist must be >0');
    if (out.cutMode !== 'planes' && out.cutMode !== 'sphere') throw new Error("--cutMode must be 'planes' or 'sphere'");
    out.minSiDegree = out.minHeavyDegree;
    if (!isFinite(out.outwardBias) || out.outwardBias < 0 || out.outwardBias > 1) throw new Error('--outwardBias must be in [0,1]');
    if (out.rcutHeavy < 0) throw new Error('--rcutHeavy must be >=0 (0=auto)');
    if (!(out.sphereR > 0)) throw new Error('--sphereR must be >0');
    if (out.sphereNrep < 0) throw new Error('--sphereNrep must be >=0');
    if (!isFinite(out.E_SiH2) || !isFinite(out.E_SiH3) || !isFinite(out.E_bare) || !isFinite(out.E_bridge) || !isFinite(out.muH)) throw new Error('Energy coefficients must be finite');

    return out;
}

async function main() {
    const args = parseArgs(process.argv);

    if (args.seed !== 0) {
        const rnd = mulberry32(args.seed);
        Math.random = rnd;
    }

    fs.mkdirSync(args.outDir, { recursive: true });

    const cifText = fs.readFileSync(args.cif, 'utf8');
    const cell = buildCrystalFromCIFText(cifText, args);
    const heavyZ = getHeavyZ(cell);

    const mm = (args.bonds || (args.caps && args.caps !== '0' && args.caps !== 'none')) ? loadMMParams(args) : null;

    const nTarget = (args.samples > 0) ? (args.samples | 0) : 0;
    const nLimit = (args.maxFiles > 0) ? (args.maxFiles | 0) : 0;
    const nOut = (nLimit > 0) ? (nTarget > 0 ? Math.min(nTarget, nLimit) : nLimit) : (nTarget > 0 ? nTarget : 0);

    if (args.samples <= 0 && args.maxFiles <= 0) {
        console.log('[gen_nanocrystals] NOTE: neither --samples nor --maxFiles specified; defaulting to 1 output');
    }
    const nGen = (nOut > 0) ? nOut : 1;

    if (args.statsCsv) {
        if (!fs.existsSync(args.statsCsv)) {
            fs.writeFileSync(args.statsCsv, 'i,nx,ny,nz,nAtoms,nBonds,nPruned,nCaps,nHHBonds,nCollapsed,nInserted,nSi,nH,nSiH,nSiH2,nSiH3,nBare,nBridge,E\n');
        }
    }

    const stackedXYZ = args.stackedXyzOut ? [] : null;

    for (let iout = 1; iout <= nGen; iout++) {
        const result = generateNanocrystal(args, cell, mm, heavyZ, iout);
        const { mol, nCollapsed, nInserted, nFused, nCaps, nPruned, nHHBonds, cnt, enr, name, nx, ny, nz } = result;

        const mol2 = toMol2String(mol, { name });
        const outPath = path.join(args.outDir, name + '.mol2');
        fs.writeFileSync(outPath, mol2);
        fs.writeFileSync(path.join(args.outDir, name + '.xyz'), toXYZString(mol, { lvec: mol.lvec }));

        if (stackedXYZ) {
            const xyz = toXYZString(mol, { lvec: mol.lvec });
            stackedXYZ.push(xyz.trimEnd());
        }

        console.log(`[gen_nanocrystals] wrote ${outPath} atoms=${mol.atoms.length} bonds=${mol.bonds.length} pruned=${nPruned} caps=${nCaps} hhBonds=${nHHBonds} collapsed=${nCollapsed} inserted=${nInserted} fused=${nFused} E=${enr.E.toFixed(6)} SiH=${cnt.nSiH} SiH2=${cnt.nSiH2} SiH3=${cnt.nSiH3} bare=${cnt.nBare}`);

        if (args.statsCsv) {
            const line = [
                iout, nx, ny, nz,
                mol.atoms.length, mol.bonds.length,
                nPruned, nCaps, nHHBonds, nCollapsed, nInserted,
                cnt.nSi, cnt.nH, cnt.nSiH, cnt.nSiH2, cnt.nSiH3, cnt.nBare, cnt.nBridge,
                enr.E
            ].join(',') + '\n';
            fs.appendFileSync(args.statsCsv, line);
        }
    }

    if (stackedXYZ) {
        const xyzPath = args.stackedXyzOut;
        fs.writeFileSync(xyzPath, stackedXYZ.join('\n'));
        console.log(`[gen_nanocrystals] wrote stacked XYZ ${xyzPath} with ${stackedXYZ.length} frames`);
    }
}

main().catch((e) => {
    console.error(e);
    process.exit(1);
});
